Add listing to cart

cheljoh opened the pull request:
Functionality for adding listing to cart implemented.

cheljoh committed:
Add listing to cart

Functionality for adding listing to cart implemented.

cheljoh committed:
Cart Count

Navbar now displays number of items in cart.

cheljoh commented:
Right now, adds all tickets in a listing to a cart (cannot choose a specific ticket)

cheljoh committed:
Add Tickets to cart

Modified cart to individual tickets are shown, not listings.

cheljoh committed:
Add Status To Tickets

Add status to tickets to the seat will only display in the listing if the status is set to active.

cheljoh committed:
house

cheljoh committed:
Sad Path

Added sad path for adding ticket to cart.
Using array instead of hash for cart contents.

cheljoh commented on app/controllers/cart_tickets_controller.rb:
```diff
@@ -0,0 +1,40 @@
+  def destroy
+    product = find_product(params[:id])
+    @cart.remove_product_from_cart(product.id.to_s)
+    flash[:alert] = "You have removed #{view_context.link_to product.name, product_path(product.id)} from your cart."
```
Going to ignore hound comment.

cheljoh commented on app/helpers/cart_helper.rb:
```diff
@@ -1,11 +1,11 @@
-    products.map do |product|
-      product.price * product.quantity
+  def raw_cart_total(tickets)
+    tickets.map do |ticket|
```
Going to ignore hound for this comment.

erinnachen commented on app/controllers/application_controller.rb:
```diff
@@ -14,4 +14,10 @@ def current_user
+      Ticket.find_by(seat: seat)
+    end
+  end
 end
```
It seems weird that this would be in application controller

cheljoh replied:
I agree, put this in cart helper!

erinnachen commented on app/helpers/cart_helper.rb:
```diff
@@ -1,11 +1,11 @@
+    tickets.map do |ticket|
+      ticket.price
     end.reduce(:+)
   end
```
you can do this in the reduce instead of having to do a map and a reduce 
it would probably be something like tickets.reduce(0){ | sum, ticket | sum+ticket.price }

erinnachen commented on spec/features/user/user_views_a_specific_event_spec.rb:
```diff
@@ -27,14 +27,29 @@
+    listing = listing1
+
+    ticket1.update(seat: "8")
+    ticket2.update(seat: "9", status: "inactive")
```
maybe the status should be "sold" or "expired"?

cheljoh committed:
Merge branch 'master' into add_ticket_to_cart

cheljoh committed:
Fix Navbar

Took unnecessary "sell tickets" link out of navbar.

cheljoh committed:
Fix seeds.rb and spec_helper

Changed above files because recent validations implemented.
User adds a listing spec does not pass right now.

cheljoh committed:
Refactor

Put "get_tickets" method in cart_helper, cleaned up spec_helper a tad.
Refactored raw_cart_total.

erinnachen commented on app/controllers/cart_tickets_controller.rb:
```diff
@@ -0,0 +1,39 @@
+    seats = params[:seats]
+    if seats.nil?
+      flash[:alert] = "please select a seat"
+      redirect_to event_path(listing.event_id)
```
should this be a render instead of a redirect?

erinnachen commented on app/helpers/cart_helper.rb:
```diff
@@ -1,11 +1,15 @@
+  def get_tickets(seats)
+    seats.map do |seat|
+      Ticket.find_by(seat: seat)
+    end
```
this is probably not going to be unique... I think it needs to be scoped to a listing: listing.tickets.find_by(seat: seat)

cheljoh replied:
Good call!

erinnachen commented on app/models/cart_ticket.rb:
```diff
@@ -0,0 +1,16 @@
+
+  def initialize(ticket_id)
+    @listing = Ticket.find(ticket_id)
+    super(@listing)
```
this feels weird... shouldn't it be `@ticket`?

cheljoh replied:
Good catch, originally added listing and forgot to change ivar.

erinnachen commented on app/models/ticket.rb:
```diff
@@ -11,4 +11,9 @@ def format_price
+
+  def get_event
+    listing = Listing.find(listing_id)
+    Event.find(listing.event_id)
```
I think you can replace this with `listing.event`

cheljoh committed:
Refactor

Scope finding tickets by listing instead of seats (which are not unique).
Render event/show instead of redirecting if form is incorrect when adding to cart.

cheljoh committed:
Change Cart Ticket and Get Event

Name variable @ticket instead of @listing for cart ticket.
Refactor get_event method.

cheljoh committed:
Fix User Can Add Listing Spec

Changed expected date in spec.

erinnachen commented on app/models/listing.rb:
```diff
@@ -52,6 +52,18 @@ def view_price
+      if ticket.status == "active"
+        ticket.seat
+      end
+    end
```
I think we can do this with an Active Record call: tickets.where(status: "active).pluck(:seat) (?) something like that...

erinnachen commented on app/models/cart.rb:
```diff
@@ -2,21 +2,22 @@ class Cart
-    contents[product_id.to_s] += quantity.to_i
+  def add_tickets(tickets)
+    tickets.map do |ticket|
+      contents << ticket.id.to_s
```
is there a reason why this is a map? also do you need to convert ticket.id to a string?

cheljoh committed:
Seeds and Seats

Change seeds file for admin.

Use Active Record query for seats instead of map.

cheljoh committed:
Refactor

Change loop from map to each. Ticket id doesn't need to be a string.

erinnachen commented:
Looks awesome!
